File: 2022/08/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

// Checks if a tree is visible.
function isVisible(grid, x, y) {
	for (let i = y + 1; i < grid.length; i++) {
		if (grid[i][x] >= grid[y][x]) break;
		if (i === grid.length - 1) return true;
	}
	for (let i = y - 1; i >= 0; i--) {
		if (grid[i][x] >= grid[y][x]) break;
		if (i === 0) return true;
	}
	for (let i = x + 1; i < grid[0].length; i++) {
		if (grid[y][i] >= grid[y][x]) break;
		if (i === grid[0].length - 1) return true;
	}
	for (let i = x - 1; i >= 0; i--) {
		if (grid[y][i] >= grid[y][x]) return false;
		if (i === 0) return true;
	}
}

// Calculates the scenic score of a tree.
function scenicScore(grid, x, y) {
	let a = 1;
	for (let i = y + 1; i < grid.length - 1; i++) {
		if (grid[i][x] >= grid[y][x]) break;
		a++;
	}
	let b = 1;
	for (let i = y - 1; i > 0; i--) {
		if (grid[i][x] >= grid[y][x]) break;
		b++;
	}
	let c = 1;
	for (let i = x + 1; i < grid[0].length - 1; i++) {
		if (grid[y][i] >= grid[y][x]) break;
		c++;
	}
	let d = 1;
	for (let i = x - 1; i > 0; i--) {
		if (grid[y][i] >= grid[y][x]) break;
		d++;
	}
	return a * b * c * d;
}

// --- Read and parse the input file ---
let grid = [];
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	grid.push(Array.from(line));
});
await new Promise(res => reader.once("close", res));

// --- Puzzle 1 ---
let total = (grid.length + grid[0].length - 2) * 2;
for (let y = 1; y < grid.length - 1; y++)
	for (let x = 1; x < grid[0].length - 1; x++)
		total += isVisible(grid, x, y) ? 1 : 0;
console.log("1. Trees that are visible from aoutside the grid: ", total.toLocaleString());

// --- Puzzle 2 ---
let max = 0;
for (let y = 1; y < grid.length - 1; y++)
	for (let x = 1; x < grid[0].length - 1; x++) {
		const score = scenicScore(grid, x, y);
		if (score > max) max = score;
	}
console.log("2. The highest scenic score is: ", max.toLocaleString());

File: 2022/08/python/main.py
# --- Constants ---
INPUT_FILE = "../input/input.txt"

# --- Checks if a tree is visible from outside ---
def is_visible(grid, x, y):
    for i in range(y + 1, len(grid)):
        if grid[i][x] >= grid[y][x]: break
        if i == len(grid) - 1: return True
    for i in reversed(range(y)):
        if grid[i][x] >= grid[y][x]: break
        if i == 0: return True
    for i in range(x + 1, len(grid[0])):
        if grid[y][i] >= grid[y][x]: break
        if i == len(grid[9]) - 1: return True
    for i in reversed(range(x)):
        if grid[y][i] >= grid[y][x]: return False
        if i == 0: return True
    raise Exception("Unreachable!")

# --- Calculates the scenic score ---
def scenic_score(grid, x, y):
    a = 1
    for i in range(y + 1, len(grid) - 1):
        if grid[i][x] >= grid[y][x]: break
        a += 1
    b = 1
    for i in reversed(range(1, y)):
        if grid[i][x] >= grid[y][x]: break
        b += 1
    c = 1
    for i in range(x + 1, len(grid[0]) - 1):
        if grid[y][i] >= grid[y][x]: break
        c += 1
    d = 1
    for i in reversed(range(1, x)):
        if grid[y][i] >= grid[y][x]: return False
        d += 1
    return a * b * c * d

# --- Read the input file ---
grid = [];

file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    grid.append([ int(x) for x in line ])

# ---- Puzzle 1 ---
total = (len(grid) + len(grid[0]) - 2) * 2
for y in range(1, len(grid) - 1):
    for x in range(1, len(grid[0]) - 1):
        if is_visible(grid, x, y): total += 1
print(f"1. Trees that are visible from outside the grid: {total:,d}")

# --- Puzzle 2 ---
score = 0
for y in range(1, len(grid) - 1):
    for x in range(1, len(grid[0]) - 1):
        score = max(score, scenic_score(grid, x, y))
print(f"2. The highest scenic score is: {score:,d}")
